### PR TITLE
POC: TS path aliases, standard tools without abstractions, hierarchical storybook

### PR DESCRIPTION
**Pull request by Hotell**
#### Description of changes

This is POC based on #16570 

- concepts were applied for `react-button` ( non converged component ), just to showcase the architecture
- going forward this architecture should be applied only on new convergence packages - as it will solve our current issues with dev loop and will properly sandbox us from the rest of the repo.

#### Architecture (simplified)

image

#### New Dev workflow (per package):

> **NOTE:** 
> - no need to `cd` into nested trees 
> - no need build on every change (even cross package related)
> - standard `help` works for every tool used
> - docs (storybook stories) are now collocated ( common practice )
> - no need for `just` nor `gulp`
> - `lage` can be used as is to parallelize builds on CI/when publishing
> - npm devDepependencies have been moved to root package.json as `single version policy` is one of the requirements (to  simplify package resolution

1. storybook
```
yarn workspace @fluentui/react-button start

# invoke help to see what's available
yarn workspace @fluentui/react-button start --help
```

2. Run tests
```
yarn workspace @fluentui/react-button test
# in watch mode
yarn workspace @fluentui/react-button test  --watch

# invoke help to see what's available
yarn workspace @fluentui/react-button test --help
```

3. Lint
```
yarn workspace @fluentui/react-button lint

# invoke help to see what's available
yarn workspace @fluentui/react-button lint --help
```

4. Format
```
yarn workspace @fluentui/react-button format

# invoke help to see what's available
yarn workspace @fluentui/react-button lint --help

# verify if you have any formatting inconsistencies
yarn workspace @fluentui/react-button format:check
```

5. Build ( only needed for publishing )

> NOTE: not implemented

```
# run 
# 1. tsc to generate declarations
# 2. run rollup to create UMD/ESM bundle (mitigating  use of any kind of import paths by customers)
yarn workspace @fluentui/react-button build
```

#### Things that are far from perfect (slow in this POC):

**Tests:**
- TS path resolutions transforms to jest config are hardcoded via relative path (we'd need to implement simple jest plugin resolver - already implemented by Nx folks )
- `isConformant` calls are extremely slow, especially on cold start - caused mainly by react-doc block

### CODE REVIEW

**Review by JustSlone — COMMENTED** [review]
Looking promising 👍 

Left some comments / questions / suggestions. I also hit a few issues and mentioned them (I know it's a DRAFT pr, so maybe these are known issues).

File: jest.preset.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    '/node_modules/',
+    '\\.pnp\\.[^\\/]+$',
+    // custom path
+    '/lib-commonjs/',
+  ],
+  //
+  testMatch: ['**/+(*.)+(spec|test).+(ts)?(x)'],
+  testURL: 'http://localhost',
+  moduleNameMapper,
+  cacheDirectory: '<rootDir>/.cache/jest',
```
Comment (ecraig12345) [inline jest.preset.js:33]: 👍  we should probably add a similar setting to our current package configuration--having jest cache in its default location outside the repo is counterintuitive

File: packages/react-button/.eslintrc.json [hunk 2]
```diff
@@ -1,4 +1,12 @@
 {
   "extends": ["plugin:@fluentui/eslint-plugin/react"],
-  "root": true
+  "root": true,
+  "overrides": [
+    {
+      "files": ["*.stories.{ts,tsx}"],
```
Comment (ecraig12345) [inline packages/react-button/.eslintrc.json:6]: If we go forward with this change or similar, this rule should move to the overrides in `@fluentui/eslint-plugin/react`.

Also a more general comment, I have a hard time understanding the value of lint rules with "warning" level. If people decide to disregard the warning, it either causes spam in logs (annoying and potentially confusing) or requires adding a disable comment (same as would happen with an error level rule that someone felt should be ignored in a particular spot). However I'm definitely not blocking on this because based on common presets it's clear that a lot of the community doesn't share my opinion.

File: packages/react-button/package.json [hunk 3]
```diff
@@ -14,26 +14,23 @@
...
-    "lint": "just-scripts lint",
-    "start": "just-scripts dev:storybook",
-    "start:legacy": "just-scripts dev",
-    "start-test": "just-scripts jest-watch",
-    "test": "just-scripts test",
-    "update-api": "just-scripts update-api",
-    "update-snapshots": "just-scripts jest -u"
+    "build": "echo 'TODO: rollup & tsc --declaration && api-extractor'",
+    "clean": "rm -rf dist lib-{amd,commonjs,es-2015} .cache",
+    "start": "yarn serve",
```
[the remark this answers is not in the document]
Comment (ecraig12345) [reply packages/react-button/package.json:19]: I suspect that isn't expected to work yet. Also a bit surprised `yarn workspace @fluentui/react-button start` even works since react-button is now excluded from the workspace config.

File: packages/react-button/package.json [hunk 4]
```diff
@@ -14,26 +14,23 @@
...
+    "serve": "start-storybook --ci -c ./.storybook -p 6006",
+    "test": "jest",
+    "lint": "eslint --ext=.ts,.tsx src/",
+    "format": "prettier **/*.{ts,tsx,md,html,scss} --write",
```
Comment (ecraig12345) [inline packages/react-button/package.json:23]: One concern I have with duplicating the configuration for each command in every package.json is that it would have a negative impact on maintainability (if the options need to be changed for some reason, as far as I can tell you'd have to rely on find/replace and hope it caught everywhere). What has been your experience with that in other projects?

File: packages/react-button/src/components/Button/Converged.stories.tsx [hunk 5]
```diff
@@ -0,0 +1,90 @@
+import React from 'react';
```
[the remark this answers is not in the document]
Comment (ecraig12345) [reply packages/react-button/src/components/Button/Converged.stories.tsx:1]: I don't think any of the stories are new files, just copied, but you bring up an interesting question.

For old packages we've used the same `*.Example.tsx` files for both development and documentation (website), but we don't appear to have a clear policy for how converged components should handle this yet. Today the "proto-converged" packages have a random mix of `*.stories.tsx` (usually exporting multiple stories) and `*.Example.tsx` (usually copied from the v8 component's examples).

More on this in a response to a later comment.

[the remark this answers is not in the document]
Comment (ecraig12345) [reply packages/react-button/src/components/Button/Converged.stories.tsx:1]: I'm not sure we need to worry too much about that right now. Like I mentioned in the previous reply, these files are just copies, so not a change to the status quo. More generally, there's nothing storybook-specific about this format aside from the name (it's just a file exporting a bunch of components), so moving to some other format shouldn't be too hard.

File: packages/react-button/src/components/Button/Converged.stories.tsx [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+  </ThemeProvider>
+);
+
+export const ThemeExample = () => {
+  const [color, setColor] = React.useState<string>('#f00');
+  const { lightTheme, darkTheme } = getThemes(color);
+  const onColorChange = React.useCallback((ev: never, newColor: IColor) => setColor(newColor.str), []);
+
+  return (
+    <Stack gap={16}>
```
[the remark this answers is not in the document]
Comment (ecraig12345) [reply packages/react-button/src/components/Button/Converged.stories.tsx:77]: Justin is right, the reason we split the stories/examples into a separate package is because having them in the component packages can create circular dependency issues. It doesn't happen in this case because react-button currently depends on react-internal, but it becomes a big issue once you remove the react-internal dep. The cases he mentioned are good examples. 

The possibilities I can think of to address the circular dep issue are: 
1. Put all the examples/stories in a single package, like we have today (downside: not ideal dev experience) 
2. For each component package, make a corresponding examples/stories package which can depend on anything (downside: each separate package adds maintenance and build overhead) 
3. Separate the concepts of dev-time examples (stories?) and documentation examples. Dev-time examples live in component packages (and can't depend on other components, at least not in circular ways) and documentation examples live in a single separate package. Historically we've used the same examples for development and documentation (IIRC in both v8 and v0) but maybe we should reconsider.

> That being said, maybe it's a crazy idea, but if we could avoid/solve the dependency problem (or maybe it's not a problem?). Then packages export their stories, which would make pulling examples from our components and community work nicely in the future. I like the idea of a component package (or set of packages) being self contained, should lead to better mix and matching of components in the future.

I think the only way to keep the stories in the same package and avoid the dependency issue would be option 3 above, and it would be dev-time stories only. But even if we did that, I don't think packages should export (or probably even publish) stories. Certainly shouldn't export them by default since that would lead to a confusing dev experience. 

This actually reminds me of another minor reason that we put the examples in a separate package: so they could theoretically be consumed in other published packages without inflating the size of the main package. In the past, when our examples lived in OUFR, they were also published with OUFR (because the website package uses the examples, and in the past the website package was published and partially reused by an internal version of the site). This led to complaints from some partners (ODSP mainly) that our package was too big. While this wasn't the main motivation for moving the examples, it didn't hurt.

I'd *like* to say that now that the website package isn't published, it would be possible to have examples in the component package (ignoring cycle concerns for the moment) and just exclude them from publishing using npmignore. However David mentioned on Friday that he may have to start publishing the website package again (something related to the module federation work??)...at which point we'd be stuck publishing everything it depends on too, including examples.

File: packages/react-hooks/package.json [hunk 7]
```diff
@@ -28,7 +28,6 @@
     "@fluentui/eslint-plugin": "^1.0.0-beta.0",
     "@types/enzyme": "3.10.3",
     "@types/enzyme-adapter-react-16": "1.0.3",
-    "@types/jest": "~24.9.0",
```
Comment (ecraig12345) [inline packages/react-hooks/package.json:31]: I think removing this may cause errors from the `import/no-extraneous-dependencies` rule.

File: tsconfig.base.json [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    "strict": true,
+    "strictFunctionTypes": false,
+    "moduleResolution": "node",
+    "jsx": "preserve",
+    "forceConsistentCasingInFileNames": true,
+    "typeRoots": ["node_modules/@types","./typings"],
+    "baseUrl": ".",
+    "rootDir": ".",
+    "paths": {
+      "@fluentui/react-button": ["packages/react-button/src/index.ts"],
```
Comment (ecraig12345) [inline tsconfig.base.json:19]: Can this be shorted to `"@fluentui/*": ["packages/*/src/index.ts"]`, with overrides for cases where that path is incorrect? `tsconfig.fluentui.json` does something similar today.

File: tsconfig.base.json [hunk 9]
```diff
@@ -0,0 +1,41 @@
+{
+  "compilerOptions": {
+    "module": "esnext",
+    "target": "ES2015",
+    "lib": ["es2017", "dom"],
```
Comment (ecraig12345) [inline tsconfig.base.json:5]: Unless we plan to start requiring consumers to use polyfills, we should only include up to es5 to ensure IE 11 support. (We most likely can't build our code itself with polyfills included due to bundle size concerns.)

File: tsconfig.base.json [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+    "esModuleInterop": true,
+    "skipLibCheck": true,
+    "skipDefaultLibCheck": true,
+    "sourceMap": true,
+    "strict": true,
+    "strictFunctionTypes": false,
+    "moduleResolution": "node",
+    "jsx": "preserve",
```
Comment (ecraig12345) [inline tsconfig.base.json:13]: Why this default?

**Review by ecraig12345 — COMMENTED** [review]
Like Justin said, this could be promising but I have some questions.

File: packages/react-button/src/components/Button/Converged.stories.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (Hotell) [reply packages/react-button/src/components/Button/Converged.stories.tsx:1]: >  I would really rather not have two separate systems for code examples.

yeah I fully agree. I just copied button stories from react-examples to be collocated. Ideally react-examples shouldn't exist, or if it does only as an e2e suite aggregator.

> how to avoid lock in with storybook stories

Heh but we are locked in already with webpack/jest/react/<name your tool here> ...

What's important is to have solid universal format for "real" docs ( md/mdx ). With that format/metadata we can switch to anything basically (what @ecraig12345  mentioned). 

_out off topic:_ from what I heard the plan is to have 1 storybook to aggregate all components out there from various repos

> so not a change to the status quo.

From my experience, consistency should be primary goal (simplicity,Dx...). With aforementioned setup, we are not consistent - eg. we have test collocated with implementation but docs live somewhere completely different . TBH I don't see any benefit in that.

File: packages/react-button/tsconfig.json [hunk 11]
```diff
@@ -1,23 +1,16 @@
...
     "noUnusedLocals": true,
-    "forceConsistentCasingInFileNames": true,
-    "strictNullChecks": true,
-    "noImplicitAny": true,
-    "moduleResolution": "node",
     "preserveConstEnums": true,
     "lib": ["es5", "dom"],
-    "typeRoots": ["../../node_modules/@types", "../../typings"],
-    "types": ["jest", "webpack-env", "custom-global"]
+    "types": ["jest","node", "webpack-env", "custom-global"]
```
[the remark this answers is not in the document]
Comment (Hotell) [reply packages/react-button/tsconfig.json:13]: that's great question actually :) I added node to be compliant with what is actually happening - in jest tests we have node env not browser.

I mentioned this in original proposal. In short any kind of polluting global scope is bad idea. 

With current setup, we are saying that any webpack-env/jest is available everywhere ( not just in test files but also within implementation).

In my proposal it is solved by having exact distinction between TS for testing/for implementation environment.

```
               tsconfig.json
              ↗       ↖
tsconfig.spec.json   tsconfig.lib.json 
```

File: packages/react-hooks/package.json (hunk 7 above)
Comment (Hotell) [reply packages/react-hooks/package.json:31]: with my proposal that lint rule should not be needed :)

File: tsconfig.base.json (hunk 10 above)
Comment (Hotell) [reply tsconfig.base.json:13]: the original proposal is about using babel for transpilation.

File: tsconfig.base.json (hunk 9 above)
Comment (Hotell) [reply tsconfig.base.json:5]: `lib` specification from TypeScript is not accurate btw - for that reason, babel should be used to properly transpile code based on browsers we want support, thus not blocking us to not be able to use latest ECMAscript spec.

> Unless we plan to start requiring consumers to use polyfills, we should only include up to es5 to ensure IE 11 support. 

Do we have somewhere proper set of requirements what is needed/ what we plan to support for convergence  ?

File: tsconfig.base.json (hunk 8 above)
Comment (Hotell) [reply tsconfig.base.json:19]: yeah it can but it doesn't conform requirements of  my proposal:
- will we able to use this for convergence only? - no

besides that
- is it explicit ? - not really

File: .storybook/main.js [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+    const tsPaths = new TsconfigPathsPlugin({
+      configFile: path.resolve(__dirname, '../tsconfig.base.json'),
+    });
+
+    const customRules = [
+      {
+        test: /\.(ts|tsx)$/,
+        use: [
+          {
+            loader: require.resolve('ts-loader'),
```
Comment (layershifter) [inline .storybook/main.js:23]: You mentioned in proposal that we will use Babel for complication, do you plan to use `babel-loader` instead? 🐱

Comment (layershifter) [reply .storybook/main.js:23]: Oops, it seems that I see the plan few lines below 💭

File: .storybook/main.js [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+          presets: [
+            ['@babel/preset-env'],
+            ['@babel/preset-typescript', { allowNamespaces: true }],
+            '@babel/preset-react',
+          ],
+          plugins: [
+            ['@babel/plugin-proposal-decorators', { legacy: true }],
+            '@babel/plugin-proposal-object-rest-spread',
+            '@babel/plugin-proposal-class-properties',
+            '@babel/plugin-syntax-dynamic-import',
```
Comment (layershifter) [inline .storybook/main.js:90]: Are you going to extract this to `.babelrc`/`babel.config.js`?

File: packages/react-button/jest.config.js [hunk 14]
```diff
@@ -1,10 +1,12 @@
...
 const path = require('path');
 
-const config = createConfig({
+module.exports = {
+  displayName: 'react-button',
+  preset: '../../jest.preset.js',
+  transform: {
+    '^.+\\.[tj]sx?$': ['babel-jest', { cwd: __dirname, configFile: './babel-jest.config.json' }],
+  },
+  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx'],
```
Comment (layershifter) [inline packages/react-button/jest.config.js:9]:
```suggestion
  transform: {
    '^.+\\.tsx?$': ['babel-jest', { cwd: __dirname, configFile: './babel-jest.config.json' }],
  },
  moduleFileExtensions: ['ts', 'tsx'],
```

nit: I don't think that we plan to we have `.js`/`.jsx` files in repo as test targets

File: packages/react-button/package.json [hunk 15]
```diff
@@ -14,26 +14,23 @@
...
-    "code-style": "just-scripts code-style",
-    "just": "just-scripts",
-    "lint": "just-scripts lint",
-    "start": "just-scripts dev:storybook",
-    "start:legacy": "just-scripts dev",
-    "start-test": "just-scripts jest-watch",
-    "test": "just-scripts test",
-    "update-api": "just-scripts update-api",
-    "update-snapshots": "just-scripts jest -u"
+    "build": "echo 'TODO: rollup & tsc --declaration && api-extractor'",
```
[the remark this answers is not in the document]
Comment (layershifter) [reply packages/react-button/package.json:17]: I like Rollup, it's awesome, I would say that it's brilliant for small libraries 💎 

_But,_ I would be careful with Rollup taking in account that Terser on consumer's side often does not perform well. If we will bundle with Rollup we will get a single file with all possible side-effects.

### Real example 1: downshift

I can dig dipper to remember, but functions in this file (https://github.com/downshift-js/downshift/blob/a13308910282c44ea2ebdcbc8a6dfee901df48a2/src/set-a11y-status.js#L26) are producing side-effects that will stay anyway in a bundle.

### Real example 2: non obvious Terser behavior

It's a recent discovery from Fluent's code, I have already shared this case with @dzearing. Constants in [this file](https://github.com/microsoft/fluentui/blob/d2c9430b1470c3d3912a436fcc86d0b6b162a9d5/packages/utilities/src/properties.ts) are well defined, what can go wrong? The first line... 💣 

```ts
const toObjectMap = (...items: (string[] | Record<string, number>)[]) => {};
// will become in JS
var toObjectMap = function () {
    var items = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        items[_i] = arguments[_i];
    }
};
```

Still looks good? But it will not be removed by Terser (_v5.5.1_) ever:

#### Input

```js
var toObjectMap = function () {
    var items = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        items[_i] = arguments[_i];
    }
};

var fooItems = toObjectMap()
```

#### Rollup 2.18.2

```js
!function(){for(var n=[],f=0;f<arguments.length;f++)n[f]=arguments[f]}();
```

#### Webpack 5.19.0

_BTW, try to find a difference 😏_ 

```js
!function(){for(var n=[],f=0;f<arguments.length;f++)n[f]=arguments[f]}();
```

File: packages/react-button/tsconfig.json (hunk 11 above)
[the remark this answers is not in the document]
Comment (JustSlone) [reply packages/react-button/tsconfig.json:13]: Got it, so if we split into spec/lib tsconfigs then we no longer need "node" in the main component config

File: packages/react-button/package.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (JustSlone) [reply packages/react-button/package.json:19]: So, does that mean once the whole dependency tree (of converged components and utilities) uses TS path aliases this will work?

Is there something else we have to do to fix the problem where storybook resolves the `"module"` field?

File: packages/react-button/package.json (hunk 15 above)
[the remark this answers is not in the document]
Comment (JustSlone) [reply packages/react-button/package.json:17]: Clearly a larger discussion. My comment is resolved as long as nothing in the PR here depends on going one way or the other.

File: packages/react-button/src/components/Button/Converged.stories.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (JustSlone) [reply packages/react-button/src/components/Button/Converged.stories.tsx:77]: Yeah, maybe I'm over complicating things with the idea of exporting stories. Perhaps we should just go with a simple setup. Or more likely this is a problem to be solved through whatever storybook setup we go with. 

I could possibly even be convinced that doc examples and dev time stories **should** be different.

File: packages/react-button/package.json (hunk 15 above)
[the remark this answers is not in the document]
Comment (Hotell) [reply packages/react-button/package.json:17]: @layershifter  interesting. well those are not constants by definitions (as it's dynamic behaviour as you pointed out) - maybe I'm missing something, but that function is needed for runtime, if it would be removed, those constants wouldn't work.

File: packages/react-button/package.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (Hotell) [reply packages/react-button/package.json:19]: > Is there something else we have to do to fix the problem where storybook resolves the "module" field?

I guess we need to remove those hardcoded fields and generate them only during build/publish (or point them to source instead of `lib`)

File: packages/react-button/src/components/Button/Converged.stories.tsx [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+      }}
+    >
+      <Text variant="xLarge">{title}</Text>
+      <Button icon={<UploadIcon />}>I am a button</Button>
+      <Button primary>I am a primary button</Button>
+    </Stack>
+  </ThemeProvider>
+);
+
+export const ThemeExample = () => {
```
[the remark this answers is not in the document]
Comment (Hotell) [reply packages/react-button/src/components/Button/Converged.stories.tsx:71]: good catch! I'm afraid this rabbit hole goes quite deep and is probably out of scope of this POC.

Why it doesn't work?

**TL;DR:**

- problem is with `React.forwardRef`, which defines our button api in incorrectly `Button` (wondering if this new config exposed a significant bug or if it ever worked for consumers ? )

**Extended explanation:**
- so `.forwardRef` has following definition:
```
function forwardRef<T, P = {}>(
    render: ForwardRefRenderFunction<T, P>
): ForwardRefExoticComponent<PropsWithoutRef<P> & RefAttributes<T>>;
```

- the issue starts here -> `PropsWithoutRef<P>` 
  - which is is mapped type defined as: 
  ```type PropsWithoutRef<P> =
        'ref' extends keyof P
            ? Pick<P, Exclude<keyof P, 'ref'>>
            : P;
   ```

- now if we pass `ButtonProps`  to `PropsWithoutRef`:
  -> `PropsWithoutRef<ButtonProps>` 
         -> we'll get following `{[x: string]: any;  [x: number]: any;}` 🚨
  -> huh what? - well `keyof ButtonProps === string | number` 🚨🚨

- in the end what we get is `ForwardRefExoticComponent<{[x: string]: any;  [x: number]: any;} & RefAttributes<HTMLElement>>` signature
  -> which results to: `(props: {🚨[x: string]: any; 🚨[x: number]: any; ref?: HTMLElement; key?: string | number}) => (ReactElement|null); `

**Summary:**💡those index types break intellisense/type checking `(props: {🚨[x: string]: any; 🚨[x: number]:any )`

File: packages/react-button/src/components/Button/Converged.stories.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (miroslavstastny) [reply packages/react-button/src/components/Button/Converged.stories.tsx:77]: +1 on the circular dependencies problem.

I think that it is better to have just one set of examples for all - dev, docs and screenshot tests.
Also as we plan to provide other teams with FUI storybook with all components, I would prefer to have just one storybook for all components instead of having per-component storybooks.
cc @levithomason as this is closely related/overlaps with his current work

File: packages/react-button/src/components/Button/Converged.stories.tsx (hunk 16 above)
[the remark this answers is not in the document]
Comment (Hotell) [reply packages/react-button/src/components/Button/Converged.stories.tsx:71]: #### Mystery solved:

The root cause of problem lies within ButtonProps definition.

While it might look all good 
```ts
export type ButtonProps = ComponentProps & React.HTMLAttributes<HTMLElement> & {...}
```

If we take closer look at `ComponentProps` we'll get 
     ->  `interface ComponentProps extends GenericDictionary {}`

going  further down the inheritance chain we found the root cause -> `GenericDictionary`

```
export type GenericDictionary = Record<string, any>;
```

using indexed keyed dictionaries creates not very solid API surface and with `any` it's very dangerous to use. In this scenario it gets even worse by how it resolves when we wanna use `keyof` operator

File: packages/react-button/src/components/Button/Converged.stories.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (Hotell) [reply packages/react-button/src/components/Button/Converged.stories.tsx:77]: **circular deps:**

First of all, thanks a lot for thorough background regarding this topic, which brings me back to "crucial need for requirements" :). if someone can handle us such a document, I guess we would save a lot of time.

- If we gonna use storybook, there is no such a thing as circular deps, as SB will traverse paths and bundle only files we define to bundle (`.stories.*` in this case) **to create an App not a package**. This fits right in with what is  @levithomason trying to achieve (SB composition from various repos).
- exporting stories as public API/package ?  This doesn't make any sense to me TBH (from storybook API point of view - your tool queries for what it needs to build )

**e2e coverage:** from my experience e2e should be completely separated from implementation
- `package + stories` -> storybook app available at url
- `package-e2e` -> separate folder/package (not publishable) contains only test suites that access storybook via url, thus testing it end to end

> Also as we plan to provide other teams with FUI storybook with all components, I would prefer to have just one storybook for all components instead of having per-component storybooks.

This covers both scenarios. 
- Ship+build all to one storybook/ship+build only per package. 
- Sidenote: from my experience it is rather annoying  if one needs to build whole suite to be able to work on 1 component.

File: packages/react-button/package.json (hunk 15 above)
[the remark this answers is not in the document]
Comment (Hotell) [reply packages/react-button/package.json:17]: discussed the issue with @layershifter offline 💪 . 

In short: There might be an issue within Terser -> property assignments are preventing tree shaking (array indexes, static props) etc

File: packages/react-button/src/components/Button/Converged.stories.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (levithomason) [reply packages/react-button/src/components/Button/Converged.stories.tsx:78]: OK, this should be simple :) Seems we all agree the "ideal" scenario is to colocate stories with components for optimal dev experience and to ship those examples to the docs for our users.

Martin, show us no circular dependencies 👍

File: packages/react-button/.eslintrc.json (hunk 2 above)
Comment (ecraig12345) [reply packages/react-button/.eslintrc.json:6]: > Note: This is just a showcase that package/domain specific configuration for linting works as expected

Got it, that makes sense.

I am actually interested in discussing general ideas of how linting ought to work some other time and considering whether our current approach is actually the right one (that was kind of why I mentioned the error thing), but agreed that it's not important to discuss now and not blocking this PR.

File: packages/react-button/package.json (hunk 4 above)
Comment (ecraig12345) [reply packages/react-button/package.json:23]: Got it, I wasn't clear on whether the manual configured commands in each package.json were intended as part of the proposal, or just a way to make things work right now (sounds like it's the latter, so I'm not too worried about it for now).

File: packages/react-hooks/package.json (hunk 7 above)
Comment (ecraig12345) [reply packages/react-hooks/package.json:31]: The lint rule would still be pretty important for preventing implicit deps in production code, but I see how it would become less important for tests.

File: tsconfig.base.json (hunk 9 above)
Comment (ecraig12345) [reply tsconfig.base.json:5]: > lib specification from TypeScript is not accurate btw - for that reason, babel should be used to properly transpile code based on browsers we want support, thus not blocking us to not be able to use latest ECMAscript spec.

What do you mean by not accurate? I know IE 11 supports a few things beyond what `lib: es5` includes, but for v7/8 it's generally been an adequate heuristic.

> > Unless we plan to start requiring consumers to use polyfills, we should only include up to es5 to ensure IE 11 support.
>
> Do we have somewhere proper set of requirements what is needed/ what we plan to support for convergence ?

This has been the perpetual headache of the past year: people from one team have one set of requirements in their minds, and the other team has a slightly different set, and often nobody notices until pretty late (when we've already wasted months of work) because nobody wrote the requirements down.

In this case, I'm not 100% clear on what the requirements are since I've only been on the edge of convergence discussions. I posted about it in the convergence channel and will try to get someone who's more closely involved to write it down.

File: packages/react-button/src/components/Button/Converged.stories.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (ecraig12345) [reply packages/react-button/src/components/Button/Converged.stories.tsx:1]: > From my experience, consistency should be primary goal (simplicity,Dx...). With aforementioned setup, we are not consistent - eg. we have test collocated with implementation but docs live somewhere completely different . TBH I don't see any benefit in that.

The only way we could have docs live in the same package is if we can find a way around the circular dependency issues mentioned elsewhere--which would be nice but seems unlikely.

File: packages/react-button/src/components/Button/Converged.stories.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (ecraig12345) [reply packages/react-button/src/components/Button/Converged.stories.tsx:78]: > * If we gonna use storybook, there is no such a thing as circular deps, as SB will traverse paths and bundle only files we define to bundle (`.stories.*` in this case) **to create an App not a package**. This fits right in with what is  @levithomason trying to achieve (SB composition from various repos).

I'm confused...how does just the fact of using storybook mean there are suddenly no circular deps? Would story files in individual component packages be excluded from the package's build and tsconfig? Where in the build process would they get type checked (important to prevent stories' API usage from getting out of date)?

I'd also appreciate more background from @levithomason about this proposal of multi-repo storybook, especially whether it's also intended to be used as the inner loop for this repo (which wouldn't make sense to me...seems way too heavy-weight for inner dev loop).

File: packages/react-button/package.json (hunk 4 above)
Comment (Hotell) [reply packages/react-button/package.json:23]: yup your latter statement. thx

File: tsconfig.base.json (hunk 9 above)
Comment (Hotell) [reply tsconfig.base.json:5]: > What do you mean by not accurate?

babel can apply transpilation set by proper browser support restrictions. In case of ie 11, `es5` is too strict and low level -> thus increasing bundle size/perf etc. If we setup browser restrictions accordingly we can  produce more concise code to be shipped to consumers.

File: packages/react-button/src/components/Button/Converged.stories.tsx (hunk 16 above)
[the remark this answers is not in the document]
Comment (JustSlone) [reply packages/react-button/src/components/Button/Converged.stories.tsx:72]: nice investigation! 🔥 

Doesn't seem like extending GenericDictionary is a good idea. Should we just fix this in button right now? (maybe in a separate PR)
